### DISCUSSION TITLE
Size of spaCy NER Custom Trained Model appears too small

### DISCUSSION THREAD

**Discussion by davesuketu215** [explosion/spaCy#8083] [Help: Coding & Implementations] [answered]
I have trained a custom spaCy NER Model with over 8.5 lakh training examples, for over 7 epochs with a batch size of 1000.
However, after performing ```nlp.to_disk(myPath)``` when I observe the model size in the directory, I can see its size as just 3.75 MBs. This encompasses the folders - ner, vocab and other files like cfg, meta.json, tokenizer.
I am training a blank model with just NER pipeline.

Am I missing out on anything while saving the model? Because the model size appears too small to me, given the humongous amount of data it is trained on. I am using spaCy 3.0.6 with the code approach and not CLI-based training.

**Comment by polm** [comment, marked as answer]
Model size shouldn't increase linearly with number of examples - features are a fixed size, and the only other increase will be things like the vocab, which will increase in size slowly. That said that does sound small. To be clear, did you try loading the model and checking its output?

> I am training a blank model with just NER pipeline.

In order for NER to work you need a tok2vec or something to convert strings to embeddings. Do you have a layer like that? If not that could be a problem. I recommend using the quickstart config, confirming it's working, and then customizing from there.

Also, I take it "lakh" is 100,000? I was not familiar with that term.

**Reply by davesuketu215** [reply]
Thank you for your answer. I have loaded my model and checked the output, it is perfectly working fine. I have also included the tok2vec component, as shown in the quickstart config. I have one final question pertaining to this thread.

I have trained my model using the SGD optimizer, batch size of 1000. However, the ```config.cfg``` generated by using ```nlp.to_disk( )``` does not contain this information. It has the default information about Adam optimizer and batcher size. Pasting the config.cfg here:

```
[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
seed = 0
gpu_allocator = null

[nlp]
lang = "en"
pipeline = ["tok2vec","ner"]
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
batch_size = 1000
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.ner]
factory = "ner"
moves = null
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy.HashEmbedCNN.v1"
pretrained_vectors = null
width = 96
depth = 4
embed_size = 2000
window_size = 1
maxout_pieces = 3
subword_features = true

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.HashEmbedCNN.v1"
pretrained_vectors = null
width = 96
depth = 4
embed_size = 2000
window_size = 1
maxout_pieces = 3
subword_features = true

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
gold_preproc = false
max_length = 0
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
gold_preproc = false
max_length = 0
limit = 0
augmenter = null

[training]
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
ents_per_type = null
ents_f = 1.0
ents_p = 0.0
ents_r = 0.0

[pretraining]

[initialize]
vectors = ${paths.vectors}
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]
```

I feel during ```nlp.to_disk( )``` not all the information is being captured. Do I need to register this information from the code somewhere, so that it reflects here? I am using the code approach and not the CLI-based training for my use-case.

**Reply by polm** [reply]
If you are training using code and not the cli training, then the information about your optimizer, epochs, and batches isn't internal to the model. You need to tell the model about it somehow in order for it to be saved in the config. 

The config is stored in `nlp.config` and is based on a dict, so if you want to you can just write values to it and they should be serialized normally, though that's not standard usage. 

On the other hand, one thing to keep in mind is that details like your optimizer and steps only make a difference at training time - they don't affect the model's inference at all, so you should have no trouble loading your saved model and using it for inference, or setting up a new training pipeline using your code or other code. The only thing you can't do is automatically infer your code-based training loop from the exported config, but since code-based training loops aren't recommended anyway it's not really a case we've planned for.
